### PR TITLE
Do not allow wrongly ordered sections

### PR DESCRIPTION
**Pull request by axic**
Screenshot 2020-02-18 at 20 33 46

### CODE REVIEW

File: lib/fizzy/parser.cpp [hunk 1]
```diff
@@ -292,9 +292,14 @@ Module parse(bytes_view input)
     input.remove_prefix(wasm_prefix.size());
 
     Module module;
+    SectionId last_id = SectionId::custom;
     for (auto it = input.begin(); it != input.end();)
     {
         const auto id = static_cast<SectionId>(*it++);
+        if (id != SectionId::custom && id <= last_id)
+            throw parser_error{"Unexpected section type"};
+        last_id = id;
```
Comment (gumb0) [inline lib/fizzy/parser.cpp]: I think you should do this assignment only if current section is not custom

Comment (axic) [reply lib/fizzy/parser.cpp]: > I think you should do this assignment only if current section is not custom

Yes, good point.

File: test/unittests/parser_test.cpp [hunk 2]
```diff
@@ -312,6 +264,13 @@ TEST(parser, type_section_functype_out_of_bounds)
     EXPECT_THROW_MESSAGE(parse(wasm), parser_error, "Unexpected EOF");
 }
 
+TEST(parser, import_section_empty)
+{
+    const auto bin = bytes{wasm_prefix} + make_section(2, make_vec({}));
+    const auto module = parse(bin);
+    EXPECT_EQ(module.typesec.size(), 0);
```
Comment (gumb0) [inline test/unittests/parser_test.cpp]: I guess you intended to check `importsec` here.
The tests below check `typesec`, too

**Comment by gumb0** [comment]
Also needs tests for incorrect order of sections.

File: test/unittests/parser_test.cpp (hunk 2 above)
Comment (axic) [reply test/unittests/parser_test.cpp]: Good catch.

**Comment by axic** [comment]
> Also needs tests for incorrect order of sections.

Yes, didn't start adding the test yet.

File: test/unittests/parser_test.cpp (hunk 2 above)
Comment (axic) [reply test/unittests/parser_test.cpp]: Moved to #194

File: lib/fizzy/parser.cpp [hunk 3]
```diff
@@ -331,9 +331,17 @@ Module parse(bytes_view input)
...
 
     Module module;
+    SectionId last_id = SectionId::custom;
     for (auto it = input.begin(); it != input.end();)
     {
         const auto id = static_cast<SectionId>(*it++);
+        if (id != SectionId::custom)
+        {
+            if (id <= last_id)
+                throw parser_error{"Unexpected out-of-order section type"};
```
Comment (chfast) [inline lib/fizzy/parser.cpp]: Because this is part of (noncritical) validation there should be option to disable this check. It should also throw `validation_error` (new exception type).

Comment (axic) [reply lib/fizzy/parser.cpp]: Well it is part of the binary format section and not the validation section, hence I believe `parser_error` is more appropriate.

Comment (chfast) [reply lib/fizzy/parser.cpp]: Ok then.

**Comment by axic** [comment]
> It looks with small modification is should be easy to also check if non-custom sections are duplicated.

It already checks for that, doesn't it?

**Comment by gumb0** [comment]
You could add a test for duplicate sections.

**Comment by axic** [comment]
Rebased. @chfast thanks for the extension. Are we OK merging this now?
